project/input/input.c

```diff
@@ -0,0 +1,290 @@
+#include "input.h"
+
+void set_ptr_null(Product *pp) {
+  pp->name = NULL;
+  pp->product_class = NULL;
+  pp->number = NULL;
+  pp->install_data = NULL;
+  pp->update_data = NULL;
+}
+
+char *get_string(FILE *ptemp, char end) {
+  char *buffer = NULL; //буфер для хранения символов
+  unsigned int capacity = 0; //размер буфера
+  unsigned int n = 0;        //сколько символов в буфере
+  int bitofchar = sizeof(char); //сколько бит в char
+  int MAX_LEN = 2048; // 1 строка максимум 4 Мб (MAX_LEN*2)
+  int c;              //сюда записываем символы
+  // получаем символы со стандартного ввода (клавиатура или файл)
+  while ((c = fgetc(ptemp)) != end && c != EOF && c != '\n' && c != ' ') {
+    // при необходимости увеличиваем размер буфера
+    if (n + 1 > capacity) {
+      if (!capacity) //если буфера нет: if (capacity == 0)
+        capacity = 32; //задаем новый, по умолчанию 32 символа
+      else if (capacity <= MAX_LEN)
+        capacity *= 2; //если размер исчерпан, удваиваем
+      else //иначе (угроза переполнения буфера):
+      {    //освобождаем память и выходим
+        free(buffer);
+        return NULL;
+      }
+
+      // меняем реальный размер буфера в памяти
+      char *temp = (char *)realloc(buffer, capacity * bitofchar);
+      if (!temp) // if (temp == NULL)
+      {
+        free(buffer);
+        return NULL;
+      }
+      buffer = temp; //перезаписываем буфер
+    }
+
+    buffer[n++] = c; //добавляем текущий символ в буфер
+  }
+
+  // если пользователь ввел пустую строку, возвращаем NULL
+  if ((n == 0 && c == EOF) || (n == 0 && c == '\n')) {
+    if (buffer)
+      free(buffer);
+    return NULL;
+  }
+
+  if (c == ' ') {
+    char c;
+    while ((c = fgetc(ptemp)) != '\n')
+      ;
```

Comment:
is_space из string.h в помощь.
Я бы оформил так
```suggestion
      while ((c = fgetc(ptemp)) != '\n') {
      }
```